### PR TITLE
Add additional composite key integrity and KDBX 4 upgrade tests

### PR DESCRIPTION
**Pull request by phoerious**
## Description

Adds additional tests for checking composite key integrity. Integrity is checked by constructing composite key variants, writing a KDBX and reading it back with the same or other variants of the composite key.

For testing challenge response components, a MockChallengeResponse class was added, which simulates a very simple challenge response device.

Special care was taken to ensure key integrity after explicit and implicit KDBX 4 upgrades.

## Motivation and context

This is to ensure that #1568 never happens again. The tests also cover all possible situations in which an explicit or implicit KDBX 4 upgrade can occur (the previous tests covered some, but not all).

## How has this been tested?

By running the tests.

## Types of changes

- ✅ Bug fix (non-breaking change which fixes an issue)

## Checklist:

- ✅ I have read the **CONTRIBUTING** document. **[REQUIRED]**
- ✅ My code follows the code style of this project. **[REQUIRED]**
- ✅ All new and existing tests passed. **[REQUIRED]**
- ✅ I have compiled and verified my code with `-DWITH_ASAN=ON`. **[REQUIRED]**

### CODE REVIEW

**Comment by droidmonkey** [comment]
Beautiful implementation, I will eval today

File: tests/TestKdbx4.cpp [hunk 1]
```diff
@@ -220,16 +322,17 @@ void TestKdbx4::testCustomData()
...
     db.setPublicCustomData(publicCustomData);
+    publicCustomData.insert("CD4", QByteArray::fromHex("ababa123ff"));
+    db.publicCustomData().insert("CD4", publicCustomData.value("CD4"));
     QCOMPARE(db.publicCustomData(), publicCustomData);
 
     const QString customDataKey1 = "CD1";
     const QString customDataKey2 = "CD2";
     const QString customData1 = "abcäöü";
     const QString customData2 = "Hello World";
     const int dataSize = customDataKey1.toUtf8().size() + customDataKey1.toUtf8().size() +
```
Comment (TheZ3ro) [inline tests/TestKdbx4.cpp]: shouldn't this be a `constexpr` ?

Comment (phoerious) [reply tests/TestKdbx4.cpp]: QString cannot be a constexpr and dataSize depends on runtime data.

Comment (TheZ3ro) [reply tests/TestKdbx4.cpp]: yeah, I was talking about dataSize. Make sense

Comment (phoerious) [reply tests/TestKdbx4.cpp]: constexpr makes most sense for global variables.

**Pull request merged**